### PR TITLE
Add tracing to OpenAI autologging

### PR DESCRIPTION
**Pull request by gabrielfu**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/gabrielfu/mlflow/pull/12267?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12267/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12267
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Add tracing to OpenAI autologging

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [X] Manual tests

Code:

```python
import mlflow
import openai

mlflow.openai.autolog(
    log_input_examples=True,
    log_model_signatures=True,
    log_models=True,
    log_datasets=True,
)
mlflow.set_experiment("openai_new_example")

client = openai.OpenAI(
    api_key=...,
)
messages = [{"role": "user", "content": "Say 'hello world!'"}]
output = client.chat.completions.create(
    model="gpt-3.5-turbo",
    messages=messages,
    temperature=0.0,
)
```

Successful call:
image

Failed call (e.g., setting temperature=5.0):
image

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Add tracing to OpenAI autologging

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

### CODE REVIEW

File: tests/openai/mock_openai.py [hunk 1]
```diff
@@ -79,6 +80,11 @@ async def chat_response_stream():
 
 @app.post("/chat/completions")
 async def chat(payload: ChatPayload):
+    if not 0.0 <= payload.temperature <= 2.0:
+        return fastapi.Response(
+            content="Temperature must be between 0.0 and 2.0",
+            status_code=400,
+        )
```
Comment (gabrielfu) [inline tests/openai/mock_openai.py:87]: this lets us test error behaviours. `openai` will throw a `openai.BadRequestError` when a 400 response is returned

**Comment by gabrielfu** [comment]
Hi @harupy , would you be able to review this? :)

File: mlflow/openai/_openai_autolog.py [hunk 2]
```diff
@@ -101,7 +114,15 @@ def patched_call(original, self, *args, **kwargs):
...
-        result = original(self, *args, **kwargs)
+        with mlflow.start_span(
+            name=self.__class__.__name__,
+            span_type=_get_span_type(self.__class__),
+        ) as span:
+            # openai does not accept positional arguments
+            # so we do not need to worry about it for now
+            span.set_inputs(kwargs)
+            result = original(self, *args, **kwargs)
+            span.set_outputs(result)
```
Comment (harupy) [inline mlflow/openai/_openai_autolog.py:125]: Can we add `log_traces` to `mlflow.openai.auotlog` as a flag to disable/enable trace logging?

File: mlflow/openai/_openai_autolog.py
Comment (harupy) [inline mlflow/openai/_openai_autolog.py:1]: Sorry for the late review, thanks for the PR!

File: mlflow/openai/_openai_autolog.py (hunk 2 above)
Comment (gabrielfu) [reply mlflow/openai/_openai_autolog.py]: sure, added!

File: mlflow/openai/_openai_autolog.py
Comment (gabrielfu) [reply mlflow/openai/_openai_autolog.py:1]: no worries :)

File: mlflow/openai/__init__.py [hunk 3]
```diff
@@ -897,6 +898,9 @@ def autolog(
             new model version of the registered model with this name.
             The registered model is created if it does not already exist.
         extra_tags: A dictionary of extra tags to set on each managed run created by autologging.
+        log_traces: If ``True``, traces are logged for OpenAI models by using
+            MlflowLangchainTracer as a callback during inference. If ``False``, no traces are
+            collected during inference. Default to ``True``.
```
Comment (harupy) [inline mlflow/openai/__init__.py:903]:
```suggestion
        log_traces: If ``True``, traces are logged for OpenAI models. If ``False``, no traces are
            collected during inference. Default to ``True``.
```

**Comment by harupy** [comment]
Manually ran `examples/openai/autologging/instantiated_client.py` on my machine. It looks great!

File: mlflow/openai/__init__.py (hunk 3 above)
Comment (gabrielfu) [reply mlflow/openai/__init__.py]: fixed

**Review by harupy — APPROVED** [review]
LGTM!

File: mlflow/openai/_openai_autolog.py [hunk 4]
```diff
@@ -75,6 +75,19 @@ def default(self, o):
...
 
+def _get_span_type(task) -> str:
+    from openai.resources.chat.completions import Completions as ChatCompletions
+    from openai.resources.completions import Completions
+    from openai.resources.embeddings import Embeddings
+
+    span_type_mapping = {
+        ChatCompletions: SpanType.CHAT_MODEL,
+        Completions: SpanType.LLM,
+        Embeddings: SpanType.EMBEDDING,
```
Comment (harupy) [inline mlflow/openai/_openai_autolog.py:86]: we might want to ignore embeddings because the output size is massive

Comment (harupy) [reply mlflow/openai/_openai_autolog.py:86]: never mind this comment :)

**Pull request merged**